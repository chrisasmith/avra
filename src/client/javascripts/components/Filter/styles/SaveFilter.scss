@import '../../../../stylesheets/modules/theme.scss';

.popup {
  width: 190px;
  padding: 16px;
}
