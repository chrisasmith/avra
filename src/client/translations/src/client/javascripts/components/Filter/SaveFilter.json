[
  {
    "id": "save_filter.save_segmentation",
    "defaultMessage": "Save Segmentation"
  },
  {
    "id": "save_filter.name",
    "defaultMessage": "Name"
  },
  {
    "id": "save_filter.cancel_btn",
    "defaultMessage": "Cancel"
  },
  {
    "id": "save_filter.save_btn",
    "defaultMessage": "Save"
  }
]