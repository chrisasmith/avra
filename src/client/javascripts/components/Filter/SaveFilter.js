import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { FormattedMessage } from 'react-intl'
import styles from './styles/SaveFilter.scss'
import TooltipPopup from '../TooltipPopup'

const propTypes = {
  onSave: PropTypes.func.isRequired,
}

class SaveFilter extends Component {
  constructor(props, context) {
    super(props, context)

    this.saveLink = React.createRef()
    this.nameInput = React.createRef()

    this.onButtonClick = this.onButtonClick.bind(this)
    this.onClosePopup = this.onClosePopup.bind(this)
    this.onNameChange = this.onNameChange.bind(this)
    this.onSave = this.onSave.bind(this)
    this.onKeyDown = this.onKeyDown.bind(this)

    this.state = {
      showPopup: false,
      name: '',
    }
  }

  onClosePopup(e) {
    if (e) { e.preventDefault() }

    this.setState({ showPopup: false, name: '' })
  }

  onButtonClick(e) {
    if (e) { e.preventDefault() }

    this.setState({ showPopup: true }, () => {
      if (this.nameInput && this.nameInput.current) {
        this.nameInput.current.focus({ preventScroll: true })
      }
    })
  }

  onNameChange(e) {
    this.setState({ name: e.target.value })
  }

  onSave(e) {
    if (e) { e.preventDefault() }

    this.props.onSave(this.state.name)
    this.onClosePopup()
  }

  onKeyDown(e) {
    if (e.keyCode === 13 && this.state.name.length > 0) {
      this.onSave()
    }
  }

  render() {
    return (
      <div className={styles.saveFilter}>
        <div>
          <a ref={this.saveLink} href="#" onClick={this.onButtonClick}>Save segmentation</a>
        </div>
        <TooltipPopup
          target={this.saveLink.current}
          show={this.state.showPopup}
          onClickOutside={this.onClosePopup}
          direction="top"
        >
          <div className={styles.popup}>
            <div className="pb-3"><b>
              <FormattedMessage
                id="save_filter.save_segmentation"
                defaultMessage="Save Segmentation"
              />
            </b></div>
            <div className="form-group">
              <label htmlFor="segmentName">
                <FormattedMessage
                  id="save_filter.name"
                  defaultMessage="Name"
                />
              </label>
              <input
                onChange={this.onNameChange}
                value={this.state.name}
                type="text"
                maxLength="25"
                className="form-control"
                id="segmentName"
                onKeyDown={this.onKeyDown}
                ref={this.nameInput}
              />
            </div>
            <div className="text-center">
              <button onClick={this.onClosePopup} type="button" className="btn btn-secondary btn-sm">
                <FormattedMessage
                  id="save_filter.cancel_btn"
                  defaultMessage="Cancel"
                />
              </button>
              <button
                disabled={this.state.name.length === 0}
                type="button"
                className="btn btn-primary btn-sm ml-3"
                onClick={this.onSave}
              >
                <FormattedMessage
                  id="save_filter.save_btn"
                  defaultMessage="Save"
                />
              </button>
            </div>
          </div>
        </TooltipPopup>
      </div>
    )
  }
}

SaveFilter.propTypes = propTypes
export default SaveFilter
