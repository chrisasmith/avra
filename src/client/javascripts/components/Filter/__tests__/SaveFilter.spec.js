import React from 'react'
import { shallow, mount } from 'enzyme'
import { shallowToJson, renderToJson } from 'enzyme-to-json'
import SaveFilter from '../SaveFilter'

jest.mock('../../TooltipPopup')

describe('Components SaveFilter', () => {
  it('should render', () => {
    const wrapper = shallow(<SaveFilter
      onSave={jest.fn()}
    />)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('should pop open', () => {
    const wrapper = mount(<SaveFilter
      onSave={jest.fn()}
    />)

    wrapper.find('a').simulate('click')
    expect(renderToJson(wrapper.render())).toMatchSnapshot()
  })

  it('should enable save', () => {
    const wrapper = mount(<SaveFilter
      onSave={jest.fn()}
    />)

    wrapper.find('a').simulate('click')
    wrapper.find('input').simulate('change', { target: { value: 'hi' } })
    expect(renderToJson(wrapper.render())).toMatchSnapshot()
  })

  it('should fire save', () => {
    const onSave = jest.fn()
    const wrapper = mount(<SaveFilter
      onSave={onSave}
    />)

    wrapper.find('a').simulate('click')
    wrapper.find('input').simulate('change', { target: { value: 'hi' } })
    wrapper.find('.btn-primary').simulate('click')
    expect(renderToJson(wrapper.render())).toMatchSnapshot()
    expect(onSave).toHaveBeenCalledWith('hi')
  })
})
