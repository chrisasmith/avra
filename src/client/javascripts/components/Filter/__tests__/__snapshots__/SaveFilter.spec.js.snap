// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components SaveFilter should enable save 1`] = `
<div
  class="saveFilter"
>
  <div>
    <a
      href="#"
    >
      Save segmentation
    </a>
  </div>
  <div>
    <div
      class="popup"
    >
      <div
        class="pb-3"
      >
        <b>
          Save Segmentation
        </b>
      </div>
      <div
        class="form-group"
      >
        <label
          for="segmentName"
        >
          Name
        </label>
        <input
          class="form-control"
          id="segmentName"
          maxlength="25"
          type="text"
          value="hi"
        />
      </div>
      <div
        class="text-center"
      >
        <button
          class="btn btn-secondary btn-sm"
          type="button"
        >
          Cancel
        </button>
        <button
          class="btn btn-primary btn-sm ml-3"
          type="button"
        >
          Save
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Components SaveFilter should fire save 1`] = `
<div
  class="saveFilter"
>
  <div>
    <a
      href="#"
    >
      Save segmentation
    </a>
  </div>
  <div>
    <div
      class="popup"
    >
      <div
        class="pb-3"
      >
        <b>
          Save Segmentation
        </b>
      </div>
      <div
        class="form-group"
      >
        <label
          for="segmentName"
        >
          Name
        </label>
        <input
          class="form-control"
          id="segmentName"
          maxlength="25"
          type="text"
          value=""
        />
      </div>
      <div
        class="text-center"
      >
        <button
          class="btn btn-secondary btn-sm"
          type="button"
        >
          Cancel
        </button>
        <button
          class="btn btn-primary btn-sm ml-3"
          disabled=""
          type="button"
        >
          Save
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Components SaveFilter should pop open 1`] = `
<div
  class="saveFilter"
>
  <div>
    <a
      href="#"
    >
      Save segmentation
    </a>
  </div>
  <div>
    <div
      class="popup"
    >
      <div
        class="pb-3"
      >
        <b>
          Save Segmentation
        </b>
      </div>
      <div
        class="form-group"
      >
        <label
          for="segmentName"
        >
          Name
        </label>
        <input
          class="form-control"
          id="segmentName"
          maxlength="25"
          type="text"
          value=""
        />
      </div>
      <div
        class="text-center"
      >
        <button
          class="btn btn-secondary btn-sm"
          type="button"
        >
          Cancel
        </button>
        <button
          class="btn btn-primary btn-sm ml-3"
          disabled=""
          type="button"
        >
          Save
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Components SaveFilter should render 1`] = `
<div
  className="saveFilter"
>
  <div>
    <a
      href="#"
      onClick={[Function]}
    >
      Save segmentation
    </a>
  </div>
  <TooltipPopup
    direction="top"
    onClickOutside={[Function]}
    show={false}
    target={null}
  >
    <div
      className="popup"
    >
      <div
        className="pb-3"
      >
        <b>
          Save Segmentation
        </b>
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="segmentName"
        >
          Name
        </label>
        <input
          className="form-control"
          id="segmentName"
          maxLength="25"
          onChange={[Function]}
          onKeyDown={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="text-center"
      >
        <button
          className="btn btn-secondary btn-sm"
          onClick={[Function]}
          type="button"
        >
          Cancel
        </button>
        <button
          className="btn btn-primary btn-sm ml-3"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          Save
        </button>
      </div>
    </div>
  </TooltipPopup>
</div>
`;
